README.md

```diff
@@ -22,6 +22,36 @@ fn main() {
 }
 ```
 
+# Safety
+
+It is generally not safe to concatenate two adjacent slices.
+This is explained in [#8], [bluss/odds#25], [rust-lang/rust#66111], [rust-lang/rust#62765], [rust-lang/rfcs#2806] and [tokio-rs/bytes#347].
+To sum it up, when rust calculates offsets into a slice, it makes the assumption that elements of that slice stem from the same allocation.
+If two different allocations are adjacent to each other, the `concat*` functions in this crate would readily combine them, breaking internal assumptions of `rustc` and thus may result in UB.
+Therefore, all functions in this crate are marked as `unsafe`.
+The user must ensure that the slices to be concatenated stem from the same allocation (the same `String`, `Vec<_>`, …).
+
+Another edge-case are zero-sized types (ZST).
+Multiple slices over different ZST may point to the same memory-region.
+That is because a ZST doesn't "exist" (i.e. has any size) during runtime.
+As such, there is no need to actually allocate any space during runtime when using a ZST (or a `Vec` or slice of it).
+Instead, a slice to ZSTs will point to the smallest non-zero address with the correct alignment (for example `vec![()].as_ptr() as usize == 1`).
+Technically it's safe to arbitrary elongate a slice of ZST given it already contains at least one element.
+However, we decided that semantically it doesn't make sense to use this functionality.
+As we also don't have any way to tell that two ZST slices are adjacent, we decided to always error if two ZST slices are tried to be concatenated.
```

Comment:
> always error when trying to concatenate two ZST slices

Otherwise we would have to organize some sort of jury duty.

Reply:
The offending part is 'are tried to' which would mean either: they are put on trial; we are trying out ZST slices themselves.